activemodel/lib/active_model/base.rb

```diff
@@ -0,0 +1,414 @@
+# frozen_string_literal: true
+
+module ActiveModel
+  # = Active Model \Base
+  #
+  # Base provides subclasses with an Active Record-inspired
+  # interface to execute code with familiar methods like +.create+, +#save+, and
+  # +#update+. It includes the API module transitively through the Model module,
+  # so it's designed to integrate with Action Pack and Action View out of the box.
+  #
+  # Similar to the convention for applications to define an
+  # <tt>ApplicationRecord</tt> that inherits <tt>ActiveRecord::Base</tt>, it's
+  # also conventional for applications to define an <tt>ApplicationModel</tt>
+  # that inherits from Base:
+  #
+  #   # app/models/application_model.rb
+  #
+  #   class ApplicationModel < ActiveModel::Base
+  #   end
+  #
+  # Unlike other facets of Active Model, Base is a Class instead of a Module.
+  # Once classes have inherited from Base, they only need to define a +#save!+.
+  # For example, consider a <tt>Session</tt> model responsible for authenticating a
+  # <tt>User</tt>  with +email+ and +password+ credentials:
+  #
+  #   # app/models/session.rb
+  #
+  #   class Session < ApplicationModel
+  #     attr_accessor :email, :password, :request
+  #
+  #     validates :email, :password, presence: true
+  #
+  #     def save!
+  #       if (user = User.authenticate_by(email: email, password: password))
+  #         request.cookies[:signed_user_id] = user.signed_id
+  #       else
+  #         errors.add(:base, :invalid)
+  #
+  #         raise ActiveModel::ValidationError.new(self)
+  #       end
+  #     end
+  #   end
+  #
+  # NOTE: This implementation is intended for demonstration purposes only, and
+  # is not meant to be used in a real application.
+  #
+  # By defining +#save!+, the +Session+ class gains access to other methods
+  # provided by Base, like +create!+ and +create+, +update!+ and +update+,
+  # +persisted?+ and +new_model?+, and all of the other utilities provided by
+  # Model, API, and Conversion.
+  class Base
```

Comment:
I had considered moving the majority of the code into an `ActiveModel::Core` module to mirror `ActiveRecord::Core`, with plans to define `ActiveModel::Base` by including that module.

It'd be more consistent, and would enable other modules and class hierarchies to utilize the code in `ActiveModel::Core` (potentially even ActiveRecord), but that would mean the existence of:

* `ActiveModel::API`
* `ActiveModel::Model`
* `ActiveModel::Core`
* `ActiveModel::Base`

Which felt like one too many different concepts with confusing similar and generic names.

Reply:
@bradgessler in case you missed this comment, with regard to https://github.com/rails/rails/pull/49647#issuecomment-1765003375.